Initial Implementation of Tasks in CMS Configurations

wddgit opened the pull request:
This includes the initial implementation of the Task feature in the Core CMS configuration code. It allows creating Tasks, adding modules to them, and associating them to sequences, paths, end paths, and the schedule. In many ways Tasks behave like Sequences except that the order modules appear on Tasks does not matter. New unit tests were added to test the Core code changes.

This PR also includes changes to migrate a lot of configuration code to use Tasks. Large parts of PAT configuration code, all PAT tools, and some parts of RECO code were migrated. In particular, code that was used by runTheMatrix.py tests, IB unit tests, and other test configurations and that was already using unscheduled execution mode was migrated. Although large numbers of configuration files were migrated, an even larger number of configuration files were not migrated yet. There remains much migration work to be done. But the unmigrated configurations are not broken by this PR and should continue to run successfully. At a future date, the other code can be incrementally migrated as resources become available to do the work.

This PR modifies the convertToUnscheduled function to work properly with the new Task feature. EDProducers and EDFilters removed from paths and endpaths are automatically placed on a Task.

There are also changes to ConfigBuilder.py.

This is the third PR submitted for this set of changes. The first 2 were closed to make significant revisions. The first  one was PR #16538 and there are many comments from Chris in that PR which have all been addressed in separate commits after the first 3 commits. The second one was PR #16925, which contains Slava's comments and a lot of discussion that followed.

wddgit committed:
Initial implementation of Task in the python configuration language

wddgit committed:
Changes to make Core unit tests work with the new Task feature

wddgit committed:
Make the runTheMatrix tests and unit tests outside the Core pass with Tasks

wddgit committed:
Add parentheses to improve code readability

wddgit committed:
Fix bugs where dumpPython was writing function calls with more than 255 arguments

wddgit committed:
Change function name to indicate it is private

wddgit committed:
Add more info to exception message

wddgit committed:
Move testTask_cff from python to python/test directory

wddgit committed:
Use OrderedSet instead of a list and a set in Task, _ModuleSequenceType, and Schedule. There should be no change in behavior.

wddgit committed:
Delete some test code unintentionally committed moments earlier

wddgit committed:
Use one DependenceOrder function for Tasks and Sequences instead of two

wddgit committed:
Rename and fix allowUnscheduled unit tests

wddgit committed:
Associate PAT task directly with Schedule instead of new PAT EndPath

wddgit committed:
Use new function to which calls both setattr and Task::add

wddgit committed:
Migrate PatAlgos tests to be consistent with last couple commits

wddgit committed:
Bug fix so last 3 commits compile

wddgit committed:
Delete extraneous lines mistakenly added that do nothing

wddgit committed:
Migrate new test recently added to CMSSW to use a Task

wddgit committed:
After rebasing on 9_0_X on 1/10/2017 there was new code that needed to be migrated to Tasks

wddgit committed:
Modify so that sequences are initialized by associating to a Task

The purpose is to eliminate the maintenance burden of maintaining
nearly duplicate lists of items in Sequences and Tasks. Note that
before this commit the modules that would be setup to run unscheduled
was intentionally identical to what it was before the Pull Request,
but after the commit this is no longer true. In places where
the existing list of unscheduled modules and the existing sequences
differed, decisions had to be made to reconcile them. (It was
my original intent for this to happen, but before this commit
I intended it as a second step in the migration after the PR)

Also rebased the entire PR again when I pushed this commit.

wddgit committed:
Some necessary additions or tests fail

wddgit committed:
Fix problems with module order that showed up in addOnTests

wddgit committed:
Fix problem revealed by a unit test.

wddgit committed:
Fix a module ordering problem in runTheMatrix 25.0 step 3

wddgit committed:
Migrate more code to use Tasks

wddgit committed:
Fix PatAlgos helpers.py unit test so it runs and passes

wddgit committed:
MassReplace functions also modify tasks associated to Schedule

wddgit committed:
Fix test configurations not run by unit tests

cmsbuild commented:
A new Pull Request was created by  @wddgit (W. David Dagenhart) for CMSSW_9_0_X.

It involves the following packages:

CommonTools/ParticleFlow
Configuration/Applications
Configuration/StandardSequences
DQMOffline/JetMET
DataFormats/FWLite
EgammaAnalysis/ElectronTools
FWCore/Concurrency
FWCore/Framework
FWCore/Integration
FWCore/ParameterSet
FWCore/PythonParameterSet
HLTrigger/Configuration
HeavyFlavorAnalysis/RecoDecay
HeavyFlavorAnalysis/Skimming
HeavyFlavorAnalysis/SpecificDecay
IOPool/Common
IOPool/Output
JetMETCorrections/Configuration
JetMETCorrections/Type1MET
PhysicsTools/Heppy
PhysicsTools/JetMCAlgos
PhysicsTools/PatAlgos
PhysicsTools/PatExamples
PhysicsTools/PatUtils
PhysicsTools/SelectorUtils
RecoBTag/CTagging
RecoBTag/Combined
RecoBTag/Configuration
RecoBTag/ImpactParameter
RecoBTag/SecondaryVertex
RecoBTag/SoftLepton
RecoEgamma/ElectronIdentification
RecoEgamma/PhotonIdentification
RecoJets/JetProducers
RecoMET/METPUSubtraction
RecoMuon/MuonIdentification
RecoMuon/MuonIsolation
RecoTauTag/Configuration
RecoTauTag/RecoTau
RecoVertex/AdaptiveVertexFinder
TopQuarkAnalysis/Configuration
TopQuarkAnalysis/Examples
TopQuarkAnalysis/TopEventProducers
TopQuarkAnalysis/TopEventSelection
TopQuarkAnalysis/TopHitFit
TopQuarkAnalysis/TopJetCombination
TopQuarkAnalysis/TopKinFitter
TopQuarkAnalysis/TopPairBSM
TopQuarkAnalysis/TopSkimming
TopQuarkAnalysis/TopTools
Validation/RecoJets
Validation/RecoMET


@perrotta, @smuzaffar, @Dr15Jones, @silviodonato, @cvuosalo, @fwyzard, @franzoni, @dmitrijus, @Martin-Grunewald, @cmsbuild, @slava77, @vanbesien, @monttj, @davidlange6 can you please review it and eventually sign? Thanks.
@ghellwig, @TaiSakuma, @gouskos, @yslai, @felicepantaleo, @abbiendi, @rappoccio, @Martin-Grunewald, @ahinzmann, @geoff-smith, @mmarionncern, @kreczko, @JyothsnaKomaragiri, @makortel, @acaudron, @jhgoh, @lgray, @jdolen, @HuguesBrun, @ferencek, @trocino, @rociovilar, @Sam-Harper, @wmtan, @GiacomoSguazzoni, @rafaellopesdesa, @rovere, @VinInn, @bellan, @nhanvtran, @gkasieczka, @jalimena, @schoef, @dgulhan, @echapon, @battibass, @calderona, @mverzett, @cbernet, @gpetruc, @imarches, @mariadalfonso, @pvmulder, @bachtis this is something you requested to watch as well.
@davidlange6, @smuzaffar you are the release manager for this.

cms-bot commands are listed here cms-sw/cmssw#13028

wddgit commented:
The following changes have been made since I closed the previous PR:

1. Made changes requested by Slava and discussed at the Core meeting a few weeks ago. Where practical, the Task is now built from a list of modules and then a Sequence is created that does nothing more than associate the Task to the Sequence.

2. Previously, test configurations broken by this PR that were in unit tests had been fixed. Now I've also fixed all the test configurations that are not in unit tests but run without a fatal exception or seg fault. Of the test configurations that I found to be already broken, I fixed the simple ones and added a comment to the more complex cases. (You can identify a test configuration needing modification by looking for the allowUnscheduled parameter set True. The parameter itself isn't the problem but configurations with the parameter are ones that will need to add and associate tasks)

3. Fixed module ordering problems that showed up in the additional tests I did.

4. Fixed some instances where modules were imported only to be cloned so they would not be attached to the process.

5. Rebased as of an IB last Friday and fixed some issues and conflicts that showed up while merging in all those changes.

The Core parts of this PR have not changed at all since it was closed.

Dr15Jones commented:
please test

Dr15Jones commented:
+1

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: f0cd92ab9ae564238a558fed2d0278030bf8dff5

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17430/17644/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test TestFWCoreIntegrationStandalone had ERRORS
---> test TestFWCorePythonFrameworkDriver had ERRORS
---> test testPythonParameterSet had ERRORS

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17430/17644/summary.html

Dr15Jones commented:
The unit test failures are not from this pull request.

Martin-Grunewald commented on HLTrigger/Configuration/python/CustomConfigs.py:
```diff
@@ -110,21 +110,13 @@ def HLTDropPrevious(process):
-    for s in process.endpaths_().keys():
-        massSearchReplaceParam(getattr(process,s),name,old,new,verbose)
+    from PhysicsTools.PatAlgos.tools.helpers import massReplaceParameter
+    massReplaceParameter(process, name, old, new, verbose)
```
What is the reason for this change?
For some reason we looped explicitly over both paths and endpaths!

wddgit commented:
Hi Martin,

First, I noticed that there were functions MassReplaceInputTag and MassReplaceParameter that looped over all modules in all Paths and EndPaths and changed either InputTags or some configuration parameter. It appeared that they were trying to modify all the modules that might run. After this PR is merged, there can be modules on a Task that is directly associated with the Schedule. Those modules can run and will not be on any Path or EndPath, so the previous loops would miss them. That is what motivated me to modify these functions. I added a loop over the Tasks associated with the Schedule to the loops over Paths and EndPaths.

Second, as I was making this change I noticed that the function MassReplaceInputTag was duplicated in two places:

https://cmssdt.cern.ch/lxr/source/Configuration/Applications/python/ConfigBuilder.py#0173
and
https://cmssdt.cern.ch/lxr/source/HLTrigger/Configuration/python/CustomConfigs.py#0112

And both already call a function from PhysicsTools/PatAlgos to do their job:

https://cmssdt.cern.ch/lxr/source/HLTrigger/Configuration/python/CustomConfigs.py#0114

It seemed logical to me to refactor and eliminate the duplication and move the function with the loops to PhysicsTools/PatAlgos. And since I did this with MassReplaceInputTag, I decided to do the same thing for MassReplaceParameter. Note that the loop over EndPaths and Paths still exists, it was just moved into a new function in PhysicsTools/PatAlgos to eliminate duplication.

If you do not like this I can do it a different way. Let me know what you want here if you want me to change it.

fwyzard commented:
Hi David,
only comment I have is that to me it would make more sense to move the function from PhysicsTools/PatAlgos to FWCore/ParameterSet .

wddgit commented:
I think moving it to FWCore/ParameterSet is a good idea. I will also have to move the other function it calls and the visitor class that uses. And at the same time there is another almost identical function that also probably also ought to be moved.

https://cmssdt.cern.ch/lxr/source/RecoHI/HiEgammaAlgos/python/HiHelperTools.py#0213

This pull request is enormous already. Unless someone objects, I would like to do this as a separate pull request after this one gets merged.

fwyzard commented:
> This pull request is enormous already. Unless someone objects, I would
like to do this as a separate pull request after this one gets merged.

Yes, of course.

wddgit commented:
The changes in this PR are now integrated into the DEVEL IB, starting with CMSSW_9_0_DEVEL_X_2017-02-08-1700. I cannot see any problems there related to the changes in this PR.

davidlange6 commented:
hi @wddgit - not surprisingly this PR needs a rebase..

wddgit commented:
I am working on the rebase. Yesterday I made the code changes affecting 4 files, but was waiting for a new IB to test with. I got sidetracked on some unrelated things today. I will probably push the fixes for the conflicting development Monday.

wddgit committed:
Fix conflicts and migrate recently merged code to Tasks

cmsbuild commented:
Pull request #17430 was updated. @perrotta, @smuzaffar, @Dr15Jones, @silviodonato, @cvuosalo, @fwyzard, @franzoni, @dmitrijus, @Martin-Grunewald, @cmsbuild, @slava77, @vanbesien, @monttj, @davidlange6 can you please check and sign again.

cmsbuild commented:
Pull request #17430 was updated. @perrotta, @smuzaffar, @Dr15Jones, @silviodonato, @cvuosalo, @fwyzard, @franzoni, @dmitrijus, @Martin-Grunewald, @cmsbuild, @slava77, @vanbesien, @monttj, @davidlange6 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 62beb98a8836a013c7ed13c5bf0b16fe65841b61

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17430/17765/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test TestFWCoreIntegrationStandalone had ERRORS
---> test TestFWCorePythonFrameworkDriver had ERRORS
---> test testPythonParameterSet had ERRORS

cmsbuild commented:
Comparison job queued.

wddgit commented:
The unit test failures are unrelated to this PR.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17430/17765/summary.html

Dr15Jones commented:
+1

Dr15Jones commented:
2017-02-14 ORP decided this pull request should not go into CMSSW_9_0 but instead wait until CMSSW_9_1_X is open.

slava77 commented:
it looks like this PR will go to 91X.
So, it's somewhat on hold.

wddgit commented:
I will continue to rebase this periodically and resolve conflicts until 91X is open. If anyone has any comments or issues, it would be nice to get them now so I could resolve them before 91X is open and this could get merged quickly into 91X when it opens. I am not aware of any unresolved issues at this point.

slava77 commented:
@wddgit 
just in case you haven't noticed already, there is a conflict in  Configuration/StandardSequences/python/Validation_cff.py

davidlange6 commented:
hold

for 91x. (but please add any concerns in the next week, this PR is in the DEVEL builds already)

cmsbuild commented:
Pull request has been put on hold by  @davidlange6
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

cmsbuild commented:
Pull request #17430 was updated. @perrotta, @smuzaffar, @Dr15Jones, @silviodonato, @cvuosalo, @fwyzard, @franzoni, @dmitrijus, @Martin-Grunewald, @cmsbuild, @slava77, @vanbesien, @monttj, @davidlange6 can you please check and sign again.

wddgit commented:
I just force pushed an update that resolved the conflict and rebased the commits in the PR on the same version as in IB 2-20-2017.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: a798b8c5b8beb7deec3ec14acbc1f233f2857012

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17430/17891/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testPythonParameterSet had ERRORS

cmsbuild commented:
Comparison job queued.

wddgit commented:
The one unit test failure is not related to this PR.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17430/17891/summary.html

cmsbuild commented:
Pull request #17430 was updated. @perrotta, @smuzaffar, @Dr15Jones, @silviodonato, @cvuosalo, @fwyzard, @franzoni, @dmitrijus, @Martin-Grunewald, @cmsbuild, @slava77, @vanbesien, @monttj, @davidlange6 can you please check and sign again.

wddgit commented:
Rebased to same commit as CMSSW_9_0_X_2017-02-26-2300

Dr15Jones commented:
@davidlange6 with pre5 out, do you have an estimate on when the `master` branch will be created?

wddgit committed:
Migrate a new test configuration to use Tasks

cmsbuild commented:
Pull request #17430 was updated. @perrotta, @smuzaffar, @Dr15Jones, @silviodonato, @cvuosalo, @fwyzard, @franzoni, @dmitrijus, @Martin-Grunewald, @cmsbuild, @slava77, @vanbesien, @monttj, @davidlange6 can you please check and sign again.

slava77 commented:
assign upgrade
just because this touches essentially all areas

cmsbuild commented:
New categories assigned: upgrade

@kpedro88 you have been requested to review this Pull request/Issue and eventually sign? Thanks

davidlange6 commented:
for 91x
